File: Apple/8.string-to-integer-atoi.java
/**
 * Algorithm:
 * Parse the fixed grammar: leading spaces, optional sign, then consecutive
 * digits. Clamp during accumulation to the signed 32-bit range.
 *
 * Java data structures:
 * long stores the partial value during parsing so overflow can be detected
 * before converting to int.
 *
 * Complexity:
 * Time O(n) in the parsed prefix, space O(1).
 */
class Solution {
    public int myAtoi(String s) {
        int i = 0;
        int n = s.length();
        while (i < n && s.charAt(i) == ' ') {
            i++;
        }
        int sign = 1;
        if (i < n && (s.charAt(i) == '+' || s.charAt(i) == '-')) {
            sign = s.charAt(i) == '-' ? -1 : 1;
            i++;
        }
        long value = 0;
        while (i < n && Character.isDigit(s.charAt(i))) {
            value = value * 10 + s.charAt(i) - '0';
            if (sign == 1 && value >= Integer.MAX_VALUE) {
                return Integer.MAX_VALUE;
            }
            if (sign == -1 && -value <= Integer.MIN_VALUE) {
                return Integer.MIN_VALUE;
            }
            i++;
        }
        return (int) (sign * value);
    }
}

File: companies/Adobe/8.string-to-integer-atoi.go
package leetcode

// MyAtoi8 parses optional spaces, optional sign, then consecutive digits. The
// value is clamped during construction to the signed 32-bit range, which is the
// important fixed-width-integer concern in Go.
//
// Time: O(n) over the parsed prefix. Space: O(1).
func MyAtoi8(s string) int {
	const intMax = 1<<31 - 1
	const intMin = -1 << 31
	i, n := 0, len(s)
	for i < n && s[i] == ' ' {
		i++
	}
	sign := 1
	if i < n && (s[i] == '+' || s[i] == '-') {
		if s[i] == '-' {
			sign = -1
		}
		i++
	}
	value := 0
	for i < n && s[i] >= '0' && s[i] <= '9' {
		value = value*10 + int(s[i]-'0')
		if sign == 1 && value >= intMax {
			return intMax
		}
		if sign == -1 && -value <= intMin {
			return intMin
		}
		i++
	}
	return sign * value
}
